File: services/orders-service/src/main/java/pl/xsware/orders/infrastructure/persistence/outbox/OutboxJpaRepository.java
package pl.xsware.orders.infrastructure.persistence.outbox;

import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.UUID;

public interface OutboxJpaRepository extends JpaRepository<OutboxMessageEntity, UUID> {

    List<OutboxMessageEntity> findByProcessedAtIsNullOrderByCreatedAtDesc(Pageable pageable);
}

File: services/payments-service/src/main/kotlin/pl/xsware/payments/infrastucture/persistence/outbox/repository/OutboxJpaRepository.kt
package pl.xsware.payments.infrastucture.persistence.outbox.repository

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import pl.xsware.payments.infrastucture.persistence.outbox.entity.OutboxEntity
import java.time.Instant
import java.util.UUID

interface OutboxJpaRepository : JpaRepository<OutboxEntity, UUID> {

    @Query(
        """
        SELECT o FROM OutboxEntity o
        WHERE o.status IN ('PENDING', 'FAILED')
          AND (o.lockedUntil IS NULL OR o.lockedUntil < :now)
        ORDER BY o.createdAt ASC
        """
    )
    fun findCandidates(@Param("now") now: Instant): List<OutboxEntity>
}
